python/ray/autoscaler/updater.py

```diff
@@ -297,6 +297,47 @@ def remote_shell_command_str(self):
             self.ssh_private_key, self.ssh_user, self.ssh_ip)
 
 
+class DockerCommandRunner(SSHCommandRunner):
+    def __init__(self, doker_config, log_prefix, node_id, provider,
+                 auth_config, cluster_name, process_runner, use_internal_ip):
+        self.ssh_command_runner = SSHCommandRunner(
+            log_prefix, node_id, provider, auth_config, cluster_name,
+            process_runner, use_internal_ip)
+        self.docker_name = doker_config["container_name"]
+        self.docker_config = doker_config
+
+    def run(self,
+            cmd,
+            timeout=120,
+            allocate_tty=False,
+            exit_on_fail=False,
+            port_forward=None,
+            with_output=False):
+        return self.ssh_command_runner.run(cmd, timeout, allocate_tty,
+                                           exit_on_fail, port_forward,
+                                           with_output)
+
+    def run_rsync_up(self, source, target):
+        self.ssh_command_runner.run_rsync_up(source, target)
+        self.ssh_command_runner.run("docker cp {} {}:{}".format(
+            target, self.docker_name, self.clean_squiggly(target)))
+
+    def run_rsync_down(self, source, target):
+        self.ssh_command_runner.run("docker cp {}:{} {}".format(
+            self.docker_name, self.clean_squiggly(source), source))
+        self.ssh_command_runner.run_rsync_down(source, target)
+
+    def remote_shell_command_str(self):
+        inner_str = self.ssh_command_runner.remote_shell_command_str().replace(
+            "ssh", "ssh -tt", 1).strip("\n")
+        return inner_str + " docker exec -it {} /bin/bash\n".format(
+            self.docker_name)
+
+    def clean_squiggly(self, string):
+        return string.replace(
+            "~", "`docker exec ray_docker env | grep HOME | cut -d'=' -f2`")
```

Comment:
I think this should only be in the front of the path. For example, if my path is "user/~file.txt", we shouldn't replace it.

Also, call it `docker_expand_user`?

Reply:
Oh--I did not realize that is a valid string. I will change it to `docker_expand_user`